base/shell/explorer/traywnd.cpp

```diff
@@ -264,6 +264,200 @@ class CStartButton
 
 };
 
+// This window class name is CONFIRMED on Win10 by WinHier.
+static const WCHAR szTrayShowDesktopButton[] = L"TrayShowDesktopButtonWClass";
+
+// The 'Show Desktop' button at edge of taskbar
+class CTrayShowDesktopButton :
+    public CWindowImpl<CTrayShowDesktopButton, CWindow, CControlWinTraits>
+{
+    LONG m_nClickedTime;
+    BOOL m_bHovering;
+    HTHEME m_hTheme;
+
+public:
+    DECLARE_WND_CLASS_EX(szTrayShowDesktopButton, CS_HREDRAW | CS_VREDRAW, COLOR_3DFACE)
+
+    CTrayShowDesktopButton() : m_nClickedTime(0), m_bHovering(FALSE)
+    {
+    }
+
+    INT WidthOrHeight() const
+    {
+#define SHOW_DESKTOP_MINIMUM_WIDTH 3
+        INT cxy = 2 * ::GetSystemMetrics(SM_CXEDGE);
+        return max(cxy, SHOW_DESKTOP_MINIMUM_WIDTH);
+    }
+
+    HRESULT DoCreate(HWND hwndParent)
+    {
+        DWORD style = WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS;
+        Create(hwndParent, NULL, NULL, style);
+        if (!m_hWnd)
+            return E_FAIL;
+
+        ::SetWindowTheme(m_hWnd, L"TaskBar", NULL);
+        return S_OK;
+    }
+
+    LRESULT OnClick(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+    {
+        // The actual action can be delayed as an expected behaviour.
+        // But a too late action is an unexpected behaviour.
+        LONG nTime0 = m_nClickedTime;
+        LONG nTime1 = ::GetMessageTime();
+        if (nTime1 - nTime0 >= 600) // Ignore after 0.6 sec
+            return 0;
+
+        // Show/Hide Desktop
+        ::SendMessageW(::GetParent(m_hWnd), WM_COMMAND, TRAYCMD_TOGGLE_DESKTOP, 0);
+        return 0;
+    }
+
+#define TSDB_CLICK (WM_USER + 100)
+
+    // This function is called from OnLButtonDown and parent.
+    VOID Click()
+    {
+        // The actual action can be delayed as an expected behaviour.
+        m_nClickedTime = ::GetMessageTime();
+        PostMessage(TSDB_CLICK, 0, 0);
+    }
+
+    LRESULT OnLButtonDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+    {
+        Click(); // Left-click
```

Comment:
On windows this responds to the button up I think

Reply:
This behavior looks quicker.

Reply:
Mark is correct, and we should stick to what Windows does.